src/main/java/org/assertj/core/api/SoftProxies.java

```diff
@@ -83,58 +106,78 @@ void collectError(Throwable error) {
   }
 
   // TODO V extends AbstractAssert ?
-  @SuppressWarnings("unchecked")
-  <V, T> V create(final Class<V> assertClass, Class<T> actualClass, T actual) {
-
+  <V, T> V create(Class<V> assertClass, Class<T> actualClass, T actual) {
     try {
-      ClassLoader classLoader = getClass().getClassLoader();
-      SimpleKey key = new SimpleKey(assertClass);
-      Class<V> proxyClass = (Class<V>) cache.findOrInsert(classLoader, key, () -> createProxy(assertClass, collector));
-
+      Class<? extends V> proxyClass = createProxy(assertClass);
       Constructor<? extends V> constructor = proxyClass.getConstructor(actualClass);
-      return constructor.newInstance(actual);
+      V instance = constructor.newInstance(actual);
+      ((AssertJProxySetup) instance).assertj$setup(new ProxifyMethodChangingTheObjectUnderTest(this), collector);
+      return instance;
     } catch (NoSuchMethodException | IllegalAccessException | InstantiationException | InvocationTargetException e) {
       throw new RuntimeException(e);
     }
   }
 
-  private <V> Class<?> createProxy(Class<V> assertClass, ErrorCollector collector) {
-
-    return byteBuddy.subclass(assertClass)
+  @SuppressWarnings("unchecked")
+  private static <V> Class<? extends V> createProxy(Class<V> assertClass) {
+    return (Class<V>) CACHE.findOrInsert(SoftProxies.class.getClassLoader(), new SimpleKey(assertClass), () -> BYTE_BUDDY.subclass(assertClass)
+                    .defineField(ProxifyMethodChangingTheObjectUnderTest.FIELD_NAME, ProxifyMethodChangingTheObjectUnderTest.class, Visibility.PRIVATE)
                     .method(METHODS_CHANGING_THE_OBJECT_UNDER_TEST)
-                    .intercept(MethodDelegation.to(new ProxifyMethodChangingTheObjectUnderTest(this)))
+                    .intercept(MethodDelegation.to(ProxifyMethodChangingTheObjectUnderTest.class))
+                    .defineField(ErrorCollector.FIELD_NAME, ErrorCollector.class, Visibility.PRIVATE)
                     .method(any().and(not(METHODS_CHANGING_THE_OBJECT_UNDER_TEST))
                                  .and(not(METHODS_NOT_TO_PROXY)))
-                    .intercept(MethodDelegation.to(collector))
+                    .intercept(MethodDelegation.to(ErrorCollector.class))
+                    .implement(AssertJProxySetup.class)
+                    .intercept(FieldAccessor.ofField(ProxifyMethodChangingTheObjectUnderTest.FIELD_NAME)
+                      .setsArgumentAt(0)
```

Comment:
what does this do exactly ? it looks like it sets a ProxifyMethodChangingTheObjectUnderTest field but where ? in the created proxy ?

Reply:
Yes, to keep the proxy classes stateless, we inject the state into instance fields. This defines a setter that defines the first argument of the `AssertJProxySetup` method to one field and the second to another one.